Add mark casualties battle step

trevan opened the pull request:
This is a part of #7823.

During the firing part of the battle, there are three steps that are performed (you can see the originals in Fire and FireAa). This is the step that marks the casualties (after the user has selected them which is another step).

## Testing
<!-- Describe any manual testing performed below. -->

## Screens Shots
<!-- If there are UI updates, include screenshots below -->

## Additional Notes to Reviewer
<!-- Add any additional details that would be helpful to reviewers -->

## Release Note

<!--
Include a release note if there is a bug fix or a visible change for players.
For format & syntax help, see:
https://github.com/triplea-game/triplea/wiki/PR-Release-Notes
-->

<!--RELEASE_NOTE--><!--END_RELEASE_NOTE-->

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/BattleActions.java:
```diff
@@ -59,4 +64,12 @@ Territory querySubmergeTerritory(
+  default Player getRemotePlayer(final GamePlayer player, final IDelegateBridge bridge) {
+    // if its the null player, return a do nothing proxy
+    if (player.isNull()) {
+      return new WeakAi(player.getName());
```
side-note, this dependency on `WeakAi` is something we really would like to get rid of:
- allows for all of weak ai to be deleted
- kinda poor design, the core game engine should not depend on AI code

trevan replied:
I grabbed this code from `AbstractBattle#getRemotePlayer`.  I'm not exactly sure why it is needed.

DanVanAtta replied:
I do recognize this code, this is one of the few and biggest dependencies on AI code from core-code IIRC. In essence we do this hack to supply a functioning player object and if my vague memory on this really holds for a default OOL algorithm. Using weakAi is a very ugly hack (which we've had for a long time; we can outright delete all of weak AI once we remove this dependency, which would be a significant simplification and removal of dead code).

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/BattleState.java:
```diff
@@ -107,4 +109,8 @@ public boolean isFirstRound() {
 
   Collection<IBattle> getDependentBattles();
+
+  @RemoveOnNextMajorRelease("Only used to handle step string changes")
```
nit, I'm not sure the annotation comment string here makes a lot of sense. Would we be just as well off if it were removed? Otherwise, can it be made more direct, is it indicating that we have to do something else beyond upgrading to a non-compatible release to remove this method?

trevan replied:
I've removed the annotation comment string.

DanVanAtta reviewed (commented):
Looks like largely existing code, just one very minor comment to look at.

trevan commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/MarkCasualties.java:
```diff
@@ -0,0 +1,190 @@
+        side.getOpposite() == DEFENSE,
+        bridge);
+
+    if (firingGroup.isSuicideOnHit()) {
```
This is where the isSuicideOnHit check is being used.  You had wondered about that in the FiringGroup PR.

DanVanAtta reviewed (approved):
Looks good, I'm assuming most of the logic is ported from an existing location.

A few merge conflicts came up, should be good to merge once resolved.

trevan commented:
> Looks good, I'm assuming most of the logic is ported from an existing location.
> 
> A few merge conflicts came up, should be good to merge once resolved.

Yes, most of the logic is ported from `Fire` and `FireAa`.  I did refactor the logic in the "select casualties" step.  I just noticed that I had missed adding my test to this PR so I've added.

I also fixed the merge conflicts.

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java:
```diff
@@ -0,0 +1,526 @@
+  }
+
+  @Test
+  void isEditMode() {
```
nit, test method feels to be missing the implication of what happens when edit mode. A `@DisplayName` perhaps should become routine for any test method over 5 lines and particularly if there is more than one assert (think of it as a way to replace a javadoc on each test method, and each one should have a summary either in a direct title or display name to describe what we are testing).

trevan replied:
I've added a display name to all of the tests

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java:
```diff
@@ -0,0 +1,526 @@
+
+    assertThat(details.getKilled().toArray(), is(targetUnits.toArray()));
+    // edit mode always sets auto calculated to true
+    assertThat(details.getAutoCalculated(), is(true));
```
nit, the comment on the assertion is essentially what the string arg for assertions is for : )

EG:
```
  assertThat("edit mode sets auto calcuated to true" , details.getAutoCalculated(), is(true));
```

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java:
```diff
@@ -0,0 +1,526 @@
+              eq(delegateBridge),
+              eq(selectCasualties),
+              (Collection<Unit>) argThat(containsInAnyOrder(targetUnits.toArray())),
+              eq(3));
```
side-note, it does look like we are doing a pretty heavy lift in this test by verifying further implementation calls. If we can have the code under test instead return a result object, then have the caller invoke the select function, it would be more testable, functional, and likely easier to understand when reading the prod code (fewer side effects)

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java:
```diff
@@ -0,0 +1,526 @@
+          unit -> {
+            final UnitAttachment unitAttachment =
+                (UnitAttachment) unit.getType().getAttachment(UNIT_ATTACHMENT_NAME);
+            when(unitAttachment.getHitPoints()).thenReturn(2);
```
It looks like there are more hit points here than hits, is the name of the test correct? "moreHitsThanHitPoints"

trevan replied:
The unit has a total of 2 hit points but has already taken one hit.  So the unit has 1 hit point left.  The amount of hits in the test is 3 and there are two units.  So there are more hits than hit points.

I think this is another terminology problem.  Is "hit points" the amount the unit has ignoring any hits they've taken or is it the amount of the user has including the hits they've taken?

DanVanAtta replied:
Pretty sure - hit points is total hit points regardless of damage. Hit points left or HP remaining is hit points minus damage.

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java:
```diff
@@ -0,0 +1,526 @@
+            when(unit.getHits()).thenReturn(1);
+          });
+
+      final FiringGroup firingGroup = new FiringGroup("", "", List.of(), targetUnits);
```
Any thoughts to create a constructor that removes the no-op arguments? EG:
```
new FiringGroup(targetUnits);
```

trevan replied:
Since this constructor is only for tests, I'd rather not.

DanVanAtta replied:
Fair enough, it's a bit disturbing though that invalid values can be injected. It would seem to put the FiringGroup into an invalid state. It makes one wonder if there is a subset of behavior that could instead be a strategy object and we could test the strategy object directly. Just my 2 cents on seeing this code in test here, I'm not sure if it's practical to fix it or not.

trevan replied:
I've updated the tests to "correctly" create a FiringGroup.  The FiringGroup constructor is now private and the only way to create them is through `FiringGroup.groupBySuicideOnHit`.

DanVanAtta reviewed (commented):
Few comments on the tests, please have a look

DanVanAtta commented:
A checkstyle cropped up:

> [ant:checkstyle] [WARN] /home/travis/build/triplea-game/triplea/game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java:381: Line is longer than 100 characters (found 103). [LineLength]
